### PR TITLE
mrustc{,-rust}: update to 1.54-20220806

### PR DESCRIPTION
**Pull request by catap**
###### Type(s)

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 12.1 21C52 x86_64
Xcode 13.2.1 13C100

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [ ] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by catap** [comment]
Well.. seems that 10.15 was killed by timeouts :(

**Comment by kencu** [comment]
BTW I tried to use the mrustc-rust to build something that uses thread_local on 10.6.8, and it failed with an error in the thread_local crate build.

I was trying to build "ripgrep".

**Comment by catap** [comment]
@kencu interesting why. I'll digg this as soon as I finished with `libcxx` ;)

File: lang/mrustc/Portfile [hunk 1]
```diff
@@ -135,12 +138,28 @@ if {${subport} eq ${name}} {
...
+        depends_lib-append \
+                path:lib/libMacportsLegacySupport.dylib:legacy-support
+
+        build.env-replace \
+                    "CXXFLAGS_EXTRA=${configure.cxxflags} ${configure.universal_cxxflags}" \
+                    "CXXFLAGS_EXTRA=${configure.cxxflags} ${configure.universal_cxxflags} -isystem${prefix}/include/LegacySupport"
+
+        build.env-replace \
+                    "LINKFLAGS_EXTRA=${configure.ldflags} ${configure.universal_ldflags}" \
+                    "LINKFLAGS_EXTRA=${configure.ldflags} ${configure.universal_ldflags} -lMacportsLegacySupport"
```
Comment (evanmiller) [inline lang/mrustc/Portfile:152]: I think the preferred method is `[legacysupport::get_library_link_flags]`

Comment (catap) [reply lang/mrustc/Portfile:152]: Yeah, and it poisons LLVM build inside mrustc-rust subport which lead to complicated to track issues.

cleaning everything after PG is much more complicated that this hacks :)

**Comment by evanmiller** [comment]
Hey, looks like mrustc 0.10 is officially out :-)

https://github.com/thepowersgang/mrustc/releases/tag/v0.10

**Comment by catap** [comment]
@evanmiller hey, I'll take a look on this :)

**Comment by evanmiller** [comment]
@catap Hmm, not sure I know how to help with that, maybe try system vs MacPorts `ar` and `ld`?

**Comment by catap** [comment]
> @catap Hmm, not sure I know how to help with that, maybe try system vs MacPorts `ar` and `ld`?

I'm a bit overloaded on another things and come back to MacPorts / mrustc as soon as I can :)

**Comment by barracuda156** [comment]
@catap Any news re `gcc` support in `mrustc`?

**Comment by catap** [comment]
> @catap Any news re `gcc` support in `mrustc`?

I’ll try to finish this one with updating to the last master of Kruste in days. Probably this week.

**Comment by barracuda156** [comment]
@catap Sounds great.

**Comment by catap** [comment]
Status on macOS 10.6:
 - `mrustc` can build a rust application which is used thread local;
 - `rustc` can't build it.

**Comment by catap** [comment]
I have an idea let me run one more round of testing it.

**Comment by barracuda156** [comment]
> I have an idea let me run one more round of testing it.

@catap Since 10.6.8 allows building and running `ppc` code, have you tried that? Or nothing gonna work with `mrustc` until `gcc` backend is properly supported?

**Comment by catap** [comment]
> > I have an idea let me run one more round of testing it.
> 
> @catap Since 10.6.8 allows building and running `ppc` code, have you tried that? Or nothing gonna work with `mrustc` until `gcc` backend is properly supported?

Well… all my patches are enabled thread local via emulated tls inside libcxx; this feature is required by rust’s std and without it support will be limited to say at least.

Maybe it is possible to reuse thread local storage from gcc, but additional patching inside rust’s std is required :)

Also, aarch64 doesn’t work because some issue inside code generation, and I expect that the same issue may happen for ppc :(

So, I plan to finish this PR without gcc backend, after that try it and if it works enable it.

**Comment by catap** [comment]
This is ready to review. Anyway, support of macOS 10.6 isn't finished yet.

1. It requires https://github.com/macports/macports-ports/pull/13995 and https://github.com/macports/macports-ports/pull/14631 to be installed automatically 
2. Anyway, if you install it by hand `mrust-rustc` port not finished on macOS 10.6;

A trivial program which is used thread local storage can't be compiled by `rustc` (but works fine with `mrustc`).

**Comment by catap** [comment]
This is almost work on macOS 10.6; the one issue remains is building / linking. Maybe it opens the next one but I really doubt that.

The issue:
 * LLVM/Clang created a symbol with initialized staitc value for ThreadLocal Storage by adding `__emutls_v`;
 * rust created a symbol name without such prefix.

As result it can't be linked because it can't find a symbol :)

A naive hack to remove this prefix leads to crash inside LLVM when it is called by rust because this kind of special symbols and they added a prefix to exclude it from a lot of logic.

So, I need to take some pause because I've spent on this last couple of days.

@mascguy do you think it is good idea to merge it as is? `mrustc` is working fine at 10.6 :)

**Comment by catap** [comment]
[Right now I've run another durty hack, which may be fixed an issue, maybe not, let see in couple of hours]

**Comment by catap** [comment]
@mascguy let's wait when for a few more days :) Maybe I'll have a time to figure it out

**Comment by catap** [comment]
Well, I successfully bootstrapped a `rustc` on macOS 10.6 via `mrustc` by `mrustc-rust` port which can compile a code:
```rust
use std::cell::RefCell;
use std::thread;

thread_local! {
    static FOO: RefCell<f32> = RefCell::new(1.0);
}

fn main() {
    FOO.with(|foo| {
        *foo.borrow_mut() = 3.0;
    });

    thread::spawn(move|| {
        FOO.with(|foo| {
            println!("inner: {}", *foo.borrow());
        });
    }).join().unwrap();

    FOO.with(|foo| {
        println!("main: {}", *foo.borrow());
    });
}
```
and that works as expected:
```sh
macos-snow-leopard:tmp catap$ env MACOSX_DEPLOYMENT_TARGET=10.6 /opt/local/libexec/mrustc-rust/bin/rustc test.rs 
macos-snow-leopard:tmp catap$ ./test 
inner: 1
main: 3
macos-snow-leopard:tmp catap$ 
```

I also broke a direct requireements for libcxx: with +emulated_tls ; anyway, it is still required by `mrustc` to be compiled because it is used `thread_local` inside its code base.

=> the first step to have a rust on macOS is done.

**Comment by catap** [comment]
@mcalhoun, I guess you may be interested in it

**Comment by catap** [comment]
@MarcusCalhoun-Lopez, I guess you may be interested in it

**Comment by catap** [comment]
Let me thread this port as ready as it can be. Support of i386, gcc/ppc and arm64 requires some future work inside mrustc

**Comment by catap** [comment]
@mascguy may I ask you to review this one?

**Review by mascguy — APPROVED** [review]
LGTM. @kencu?

**Comment by mascguy** [comment]
> LGTM. @kencu?

Ken, let's take advantage of the idle buildbots, and release this into the wild. We can always submit a subsequent PR, if any changes are necessary.

Great stuff Kirill!

**Comment by mascguy** [comment]
Kirill, I'm guessing you're probably monitoring the buildbots anyway. But just to confirm, should these ports build on 10.7?

https://build.macports.org/builders/ports-10.7_x86_64-builder/builds/108738/steps/install-port/logs/stdio

**Comment by mascguy** [comment]
> Kirill, I'm guessing you're probably monitoring the buildbots anyway. But just to confirm, should these ports build on 10.7?
> 
> https://build.macports.org/builders/ports-10.7_x86_64-builder/builds/108738/steps/install-port/logs/stdio

Interestingly enough, the build for `mrustc` succeeded on 10.6 and 10.8 though.

**Comment by catap** [comment]
> > Kirill, I'm guessing you're probably monitoring the buildbots anyway. But just to confirm, should these ports build on 10.7?
> > https://build.macports.org/builders/ports-10.7_x86_64-builder/builds/108738/steps/install-port/logs/stdio
> 
> Interestingly enough, the build for `mrustc` succeeded on 10.6 and 10.8 though.

Christopher, indeed. I’ll take a look in couple hours

**Comment by mascguy** [comment]
> > > Kirill, I'm guessing you're probably monitoring the buildbots anyway. But just to confirm, should these ports build on 10.7?
> > > https://build.macports.org/builders/ports-10.7_x86_64-builder/builds/108738/steps/install-port/logs/stdio
> > 
> > 
> > Interestingly enough, the build for `mrustc` succeeded on 10.6 and 10.8 though.
> 
> Christopher, indeed. I’ll take a look in couple hours

Not an emergency, as builds look fine everywhere else. Thanks Kirill!

**Comment by catap** [comment]
@mascguy reported it to upstream:
 - https://github.com/thepowersgang/mrustc/issues/284
 - https://github.com/thepowersgang/mrustc/issues/285
 
 It can be a real bug inside mrustc :)

**Comment by catap** [comment]
Status update: the 10.7 issue was tracked and a fix was suggested: https://github.com/thepowersgang/mrustc/pull/286

**Pull request merged**